samza-core/src/main/java/org/apache/samza/execution/PlanJsonGenerator.java

```diff
@@ -0,0 +1,295 @@
+    private int opId;
+    private int pairNodeId = -1;  //for join operator, we will have a pair nodes for two partial joins
+
+    public int getPairNodeId() {
+      return pairNodeId;
+    }
+
+    public void setPairNodeId(int pairNodeId) {
+      this.pairNodeId = pairNodeId;
+    }
+
+    public String getOpCode() {
+      return opCode;
+    }
+
+    public void setOpCode(String opCode) {
+      this.opCode = opCode;
+    }
+
+    public int getOpId() {
+      return opId;
+    }
+
+    public void setOpId(int opId) {
+      this.opId = opId;
+    }
+  }
+
+  static final class StreamNode extends GraphNode {
+    private String streamId;
+    private String system;
+    private String physicalName;
+    private int partition;
+
+    public String getStreamId() {
+      return streamId;
+    }
+
+    public void setStreamId(String streamId) {
+      this.streamId = streamId;
+    }
+
+    public String getSystem() {
+      return system;
+    }
+
+    public void setSystem(String system) {
+      this.system = system;
+    }
+
+    public String getPhysicalName() {
+      return physicalName;
+    }
+
+    public void setPhysicalName(String stream) {
+      this.physicalName = stream;
+    }
+
+    public int getPartition() {
+      return partition;
+    }
+
+    public void setPartition(int partition) {
+      this.partition = partition;
+    }
+  }
+
+  static final class Job {
+    private String jobName;
+    private String jobId;
+    private List<OpNode> opNodes;
+
+    public String getJobName() {
+      return jobName;
+    }
+
+    public void setJobName(String jobName) {
+      this.jobName = jobName;
+    }
+
+    public String getJobId() {
+      return jobId;
+    }
+
+    public void setJobId(String jobId) {
+      this.jobId = jobId;
+    }
+
+    public List<OpNode> getOpNodes() {
+      return opNodes;
+    }
+
+    public void setOpNodes(List<OpNode> opNodes) {
+      this.opNodes = opNodes;
+    }
+  }
+
+  static final class GraphNodes {
+    private List<Job> jobs;
+    private List<StreamNode> streams;
+
+    public List<Job> getJobs() {
+      return jobs;
+    }
+
+    public void setJobs(List<Job> jobs) {
+      this.jobs = jobs;
+    }
+
+    public List<StreamNode> getStreams() {
+      return streams;
+    }
+
+    public void setStreams(List<StreamNode> streams) {
+      this.streams = streams;
+    }
+  }
+
+  private int nextID = 0;
+  // streamId to StreamNode mapping
+  private final Map<String, StreamNode> streamNodes = new HashMap<>();
+  // Mapping from the output stream to the join spec. Since StreamGraph creates two partial join operators for a join and they
+  // will have the same output stream, this mapping is used to choose one of them as the unique join spec representing this join
+  // (who register first in the map wins).
+  Map<MessageStream, OperatorSpec> outputStreamToJoinSpec = new HashMap<>();
+
+  /* package private */ String toJson(JobGraph jobGraph) throws Exception {
```

Comment:
Looks like this class is stateful and is meant to be created once and saved for reuse. Is this method idempotent?